include/dgl/vector.h

```diff
@@ -0,0 +1,208 @@
+    this->arr = static_cast<T *>(malloc(size * sizeof(T)));
+    this->capacity = size;
+    this->curr = 0;
+    this->own = true;
+  }
+
+  ~vector() {
+    // If the memory is allocated by the vector, it should be free'd.
+    if (this->own) {
+      free(this->arr);
+    }
+  }
+
+  vector(const vector &other) = delete;
+
+  /*
+   * Initialize the vector whose memory is allocated outside.
+   * There are no elements in the vector.
+   */
+  void init(T *arr, size_t size) {
+    CHECK(this->arr == nullptr);
+    this->arr = arr;
+    this->capacity = size;
+    this->curr = 0;
+    this->own = false;
+  }
+
+  /*
+   * Initialize the vector whose memory is allocated outside.
+   * There are elements in the vector.
+   */
+  void init(T *arr, size_t capacity, size_t size) {
+    CHECK(this->arr == nullptr);
+    CHECK_LE(size, capacity);
+    this->arr = arr;
+    this->capacity = capacity;
+    this->curr = size;
+    this->own = false;
+  }
+
+  /* Similar to std::vector::push_back. */
+  void push_back(T val) {
+    // If the vector doesn't own the memory, it can't adjust its memory size.
+    if (!this->own) {
+      CHECK_LT(curr, capacity);
+    } else if (curr == capacity) {
+      this->capacity = this->capacity * 2;
+      this->arr = static_cast<T *>(realloc(this->arr, this->capacity * sizeof(T)));
+      CHECK(this->arr) << "can't allocate memory for a larger vector.";
+    }
+    this->arr[curr++] = val;
+  }
+
+  /*
+   * This inserts multiple elements to the back of the vector.
+   */
+  void insert_back(const T* val, size_t len) {
+    if (!this->own) {
+      CHECK_LE(curr + len, capacity);
+    } else if (curr + len > capacity) {
+      this->capacity = curr + len;
+      this->arr = static_cast<T *>(realloc(this->arr, this->capacity * sizeof(T)));
+      CHECK(this->arr) << "can't allocate memory for a larger vector.";
+    }
+    std::copy(val, val + len, this->arr + curr);
+    curr += len;
+  }
+
+  /*
+   * Similar to std::vector::[].
+   * It checks the boundary of the vector.
+   */
+  T &operator[](size_t idx) {
+    CHECK_LT(idx, curr);
+    return this->arr[idx];
+  }
+
+  /*
+   * Similar to std::vector::[].
+   * It checks the boundary of the vector.
+   */
+  const T &operator[](size_t idx) const {
+    CHECK_LT(idx, curr);
+    return this->arr[idx];
+  }
+
+  /* Similar to std::vector::size. */
+  size_t size() const {
+    return this->curr;
+  }
+
+  /* Similar to std::vector::resize. */
+  void resize(size_t new_size) {
+    if (!this->own) {
+      CHECK_LE(new_size, capacity);
+    } else if (new_size > capacity) {
+      this->capacity = new_size;
+      this->arr = static_cast<T *>(realloc(this->arr, this->capacity * sizeof(T)));
+      CHECK(this->arr) << "can't allocate memory for a larger vector.";
+    }
+    for (size_t i = this->curr; i < new_size; i++)
+    this->arr[i] = 0;
+    this->curr = new_size;
+  }
+
+  /* Similar to std::vector::clear. */
+  void clear() {
+    this->curr = 0;
+  }
```

Comment:
this semantic is the same as std::vector.

Reply:
No, `std::vector::clear` will call the destructor of each element stored.

There are also couple of other details. For example, the `push_back` will fail if the size is larger than the capacity but it is not the case for `std::vector`. To avoid confusion, it's better to simply expose APIs for direct memory access such as  `operator []` or `data`. Checkout [`array_view`](https://github.com/dmlc/dmlc-core/blob/master/include/dmlc/array_view.h) from `dmlc-core`. I feel the case is quite similar. The difference is that we need non-const `operator []`, `begin` and `end`.

Are you planning to use the vector together with some other libraries that require `std::vector` interfaces? This would be a strong motivation to have a fully-compatible vector class.

Reply:
I think a better option is to put `vector` inside `CSR`. after all, `vector` is specifically designed for `CSR`. It doesn't mean to be general-purpose.

Reply:
It was implemented so until @BarclayII suggested moving it out. Inside `CSR`, it's just a helper class to implement CSR.

Reply:
I see.  I got confused thinking that this `vector` is meant to be general-purpose.